server/src/main/java/org/opensearch/search/aggregations/bucket/terms/StreamNumericTermsAggregator.java

```diff
@@ -0,0 +1,542 @@
+import org.opensearch.search.aggregations.AggregatorFactories;
+import org.opensearch.search.aggregations.BucketOrder;
+import org.opensearch.search.aggregations.CardinalityUpperBound;
+import org.opensearch.search.aggregations.InternalAggregation;
+import org.opensearch.search.aggregations.InternalMultiBucketAggregation;
+import org.opensearch.search.aggregations.InternalOrder;
+import org.opensearch.search.aggregations.LeafBucketCollector;
+import org.opensearch.search.aggregations.LeafBucketCollectorBase;
+import org.opensearch.search.aggregations.bucket.LocalBucketCountThresholds;
+import org.opensearch.search.aggregations.support.ValuesSource;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static java.util.Collections.emptyList;
+import static org.opensearch.search.aggregations.InternalOrder.isKeyOrder;
+
+/**
+ * Aggregate all docs that contain numeric terms through streaming
+ *
+ * @opensearch.internal
+ */
+public class StreamNumericTermsAggregator extends TermsAggregator {
+    private final ResultStrategy<?, ?> resultStrategy;
+    private final ValuesSource.Numeric valuesSource;
+    private final IncludeExclude.LongFilter longFilter;
+    private LongKeyedBucketOrds bucketOrds;
+    private final CardinalityUpperBound cardinality;
+
+    public StreamNumericTermsAggregator(
+        String name,
+        AggregatorFactories factories,
+        Function<StreamNumericTermsAggregator, ResultStrategy<?, ?>> resultStrategy,
+        ValuesSource.Numeric valuesSource,
+        DocValueFormat format,
+        BucketOrder order,
+        BucketCountThresholds bucketCountThresholds,
+        SearchContext aggregationContext,
+        Aggregator parent,
+        SubAggCollectionMode subAggCollectMode,
+        IncludeExclude.LongFilter longFilter,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, aggregationContext, parent, bucketCountThresholds, order, format, subAggCollectMode, metadata);
+        this.resultStrategy = resultStrategy.apply(this); // ResultStrategy needs a reference to the Aggregator to do its job.
+        this.valuesSource = valuesSource;
+        this.longFilter = longFilter;
+        this.cardinality = cardinality;
+    }
+
+    @Override
+    public void doReset() {
+        super.doReset();
+        Releasables.close(bucketOrds);
+        bucketOrds = null;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (bucketOrds != null) {
+            bucketOrds.close();
+        }
+        bucketOrds = LongKeyedBucketOrds.build(context.bigArrays(), cardinality);
```

Comment:
what is the cardinality here? Ideally, it should be bound by current leaf's max cardinality for the field.

Reply:
it just determines if we are going to collect from single bucket or multiple buckets